docs/proposals/fp2.rst

```diff
@@ -0,0 +1,145 @@
+"global rules" here for sake of clarity.
+
+However, in many cases, it is also useful for projects using Fixit to build
+custom lint rules specific to the project being linted. It is beneficial for
+those rules to be defined within the codebase being linted, without needing to
+first install those lint rules into the environment before running Fixit.
+
+These are defined as "local rules", and will require dedicated behavior from
+Fixit to discover and import these rules at runtime, as well as special
+configuration syntax for enabling these local rules in a project.
+
+
+Configuration
+-------------
+
+When configuring the set of enabled or disabled rules, any local rules must
+be marked with a single leading period, and referred to using their path
+relative to the configuration file's directory.
+
+For example, a configuration file at ``project/fixit.toml`` could include
+rules defined in ``project/some/local/rules.py`` with the following:
+
+.. code-block:: toml
+
+    [tool.fixit]
+    enable = [".some.local.rules"]
+
+
+References to local rules should be accepted in either ``enable`` or ``disable``
+options, including overrides, to provide the same selection criteria available
+to global rules:
+
+.. code-block:: toml
+
+    [tool.fixit]
+    enable = [".rules"]
+
+    [[tool.fixit.overrides]]
+    path = "project1"
+    disable = [".rules.PickyRule"]
+    enable = [".project1.rules"]
+
+
+Implementation
+--------------
+
+When gathering enabled/disabled rules from configuration, Fixit currently
+just records the module name (fqdn) and does set operations when merging
+configs/overrides. In order to differentiate and manage local rules correctly,
+without affecting the behavior of merging/overrides, one possible option is to
+replace the simple string with a tuple of path and module.
+
+For example, where a global rule would be tracked as just ``"fixit.rules"``,
+local rules from ``foo/bar/fixit.toml`` could be tracked as
+``(Path('foo/bar'), ".local.rules"))``, allowing set operations while still
+tracking their origin, and preventing collisions from different directories.
+
+In this implementation, the path object should match the parent directory
+containing the configuration file that enables or disables these local rules.
+Options in both a ``pyproject.toml`` and ``fixit.toml`` from the same exact
+directory would use the same path object, as would any local rules referenced
+by overrides in those files.
+
+---
+
+When discovering and loading rules, the system should attempt to make sure
+that it is loading rules from the local path, rather than accidentally loading
+rules from the outside environment. This can be done either by a temporarily
+restricted path when importing, or with a custom import loader.
+
+The former may be a simpler starting point, and can be handled with a temporary
+override of ``sys.path`` to include the local directory — and nothing else? —
```

Comment:
For a prototype, this is fine, but we'll need to find a better way.

1. you can't import any name that has already been imported (e.g. `enable = [".sys"]` will not work
2. if two levels have the same name (like ".lints" let's say) they will collide and be reused
3. You won't be able to import outside of that dir (this might be on purpose, but will encourage them to do sys.path hackery in the linters, and isn't explicitly documented here)

I don't see anything in importlib with custom loaders that does exactly what we'd want, so I will look in e.g. pytest or something else that imports plugins to see if there is some API I'm missing.

Reply:
And 4. rules won't be able to use relative imports within themselves.

Item 4 specifically can be fixed with [importlib.import_module](https://docs.python.org/3/library/importlib.html#importlib.import_module) specifying `package` as the one containing the toml file.  I don't _think_ this will execute intermediate `__init__.py` (which might be problematic in a megarepo), and helps with 1 and 2 but maybe makes 3 make more sense but encourage less explicit dependencies.